### PR TITLE
Each Symbol keeps a list of SymExprs

### PR DESCRIPTION
**Pull request by mppf**
This PR enables each Symbol to have a list of SymExprs that refer to it.
These could be called "uses" but in compiler terminology, a "use" is a read of a value and a "def" is a write to it. The list of Symbols that refer to a SymExprs include both "defs" and "uses" as well as other ways that a Symbol might be referred to.

For example, once resolution runs, there are SymExprs that refer to:
 *  a called function Symbol
 * a particular field in an AggregateType

Having easy access to the list of SymExprs that refer to a given Symbol will simplify some code patterns.

The list of SymExprs for each Symbol is actually implemented as a doubly-linked-list. That has the best complexity. It offers O(1) performance for adding and removing and O(n) for traversal, which is pretty much the best we could hope for. Also, these list elements are implemented with intrusive list pointers. I manually re-instantiated some other doubly-linked list code we had in the compiler for alist. I think it's important to store the list pointers within the SymExprs in order to minimize allocation overhead.

This PR is best viewed as a series of commits:

 1) Basic AST structure for Symbol listing SymExprs (ie add fields to Symbol and SymExpr)
 2) Replace all uses of SymExpr->var with SymExpr->symbol() or SymExpr->setSymbol()
 3) Wire up the list of SymExprs for each Symbol so that it updated appropriately as SymExprs are added to the tree or when setSymbol is used to update a symbol already in the tree. Also update --verify to check that this list is correct.
 4) Demonstrate the usefulness of this approach by modifying 3 loops that previously went over all SymExprs (with gSymExprs). As part of these changes, I was able to simplify these loops by removing uses of the internal Vec-set and replacing them with std::vector.

By far the majority of changes are in part (2), so it's best to view that commit separately.

Performance Impact:

I observed about 0.1 s increase in compilation time when building hello.chpl on my system before updating any passes to use the new list. Thus, the overhead is small enough that we expect performance benefit from using this approach in more of the compiler.

Future Work:

In the future, I hope that some common patterns such as finding defs and uses and finding callers of a FnSymbol will move over to the new strategy.

Testing:

Passed quickstart testing.
Passed quickstart+gasnet testing.
Passed default+verify testing.
Passed default+baseline testing.
Passed hellos with quickstart+gasnet.

Review:

Reviewed by @lydia-duncan - thanks!

### CODE REVIEW

File: compiler/AST/symbol.cpp [hunk 1]
```diff
@@ -290,6 +298,51 @@ bool Symbol::noDocGen() const {
 }
 
 
+void Symbol::addSymExpr(SymExpr* se) {
+
+  // MPF 2016-11-08: Consider not tracking SymExprs
+  // that refer Symbols that have an immediate.
```
Comment (lydia-duncan) [inline compiler/AST/symbol.cpp]: "refer Symbols" -> "refer to Symbols"

File: compiler/AST/symbol.cpp [hunk 2]
```diff
@@ -290,6 +298,51 @@ bool Symbol::noDocGen() const {
...
+  // The reason is that immediates are usually
+  // unique-ifiied and so the symbol for 3 would refer
```
Comment (lydia-duncan) [inline compiler/AST/symbol.cpp]: Double "i" in "unique-ifiied"

File: compiler/AST/symbol.cpp [hunk 3]
```diff
@@ -290,6 +298,51 @@ bool Symbol::noDocGen() const {
...
+  // to all the uses of 3, and that probably isn't adding
+  // any value.
```
Comment (lydia-duncan) [inline compiler/AST/symbol.cpp]: Conceptually, I think that's a good thing to think about, but I'm not sure the complexity add would be worthwhile (which is I suspect why you didn't do it)

File: compiler/AST/symbol.cpp [hunk 4]
```diff
@@ -3515,6 +3568,9 @@ VarSymbol* new_BoolSymbol(bool b, IF1_bool_type size) {
   imm.const_kind = NUM_KIND_BOOL;
   imm.num_index = size;
   VarSymbol *s;
+  // doesn't use uniqueConstantsHash because new_BoolSymbol is only
+  // called to initialize dtBools[i]->defaultValue.
+  // gTrue and gFalse are set up directly in initPrimitiveTypes.
```
Comment (lydia-duncan) [inline compiler/AST/symbol.cpp]: Accidental commit?  This doesn't seem particularly related at the moment.

File: compiler/optimizations/removeEmptyRecords.cpp [hunk 5]
```diff
@@ -31,7 +32,7 @@ removeEmptyRecords() {
 
   Vec<Type*> emptyRecordTypeSet;
   int numEmptyRecordTypes = 0;
-  Vec<Symbol*> emptyRecordSymbolSet;
+  std::vector<Symbol*> emptyRecordSymbols;
```
Comment (lydia-duncan) [inline compiler/optimizations/removeEmptyRecords.cpp]: :+1:

**Comment by lydia-duncan** [comment]
This looks great, thanks!  I would also mention that as part of the work you replaced a couple of internally developed STL containers with the proper library ones.  That was really nice!

File: compiler/AST/symbol.cpp (hunk 4 above)
Comment (mppf) [inline compiler/AST/symbol.cpp]: I looked at the code around how immediate Symbols work when I was considering not tracking them. At first it confused me why new_BoolSymbol doesn't have the same general pattern as the others. Hence this comment. It is meant to be part of this PR, but it's just a comment...

**Pull request merged**